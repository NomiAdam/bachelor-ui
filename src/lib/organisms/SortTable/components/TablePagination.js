import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import {
    inc, dec, lt, gte, and,
} from 'ramda';
import {
    FaAngleDoubleLeft,
    FaAngleDoubleRight,
    FaAngleLeft,
    FaAngleRight,
} from 'react-icons/lib/fa';
import styled from 'styled-components';
import { basicRowCount } from './constants/index';
import SelectField from '../../../molecules/SelectField/index';
import { darkTheme, basicTheme } from '../../../constants/theme';
import { Grid, GridCol } from '../../../atoms/Grid/index';

const StyledDiv = styled.div`
    position: relative;
    width: 100%;
    display: flex;
    justify-content: center;
    align-items: center;
    padding: 15px 0;
    font-size: .9em;
    color: ${ darkTheme.black };
    margin: 0 10px;
`;

const Icon = styled.div`
    padding: 0 10px;
    font-size: 2.8rem;
    transition: all 1s ease;
    cursor: ${ ({ disabled }) => (disabled ? 'not-allowed' : 'pointer') };
    color: ${ ({ disabled }) => (disabled ? basicTheme.grey : darkTheme.black) };
`;

const StyledSpan = styled.span`
  padding: 0 20px;
`;

const StyledFaAngleDoubleLeft = Icon.withComponent(FaAngleDoubleLeft);
const StyledFaAngleDoubleRight = Icon.withComponent(FaAngleDoubleRight);
const StyledAngleLeft = Icon.withComponent(FaAngleLeft);
const StyledAngleRight = Icon.withComponent(FaAngleRight);

const greaterThanZero = lt(0);

class TablePagination extends PureComponent {
    componentDidUpdate() {
        const {
            totalRows, rowsPerPage, currentPage, handleChangePage,
        } = this.props;
        const totalPages = Math.ceil(totalRows / rowsPerPage);
        if (currentPage > totalPages) {
            handleChangePage(totalPages);
        }
        if (totalPages > 0 && currentPage < 1) {
            handleChangePage(1);
        }
        if (totalPages === 0) {
            handleChangePage(0);
        }
    }

    render() {
        const {
            totalRows, currentPage, handleChangeRowsPerPage, handleChangePage, rowsPerPage,
        } = this.props;

        const totalPages = Math.ceil(totalRows / rowsPerPage);

        const handleChange = value => () => {
            const lessOrEqualToTotal = gte(totalPages);
            const validPageValue = and(greaterThanZero(value), lessOrEqualToTotal(value));
            if (validPageValue) handleChangePage(value);
        };
        const isFirst = currentPage === 1;
        const isLast = currentPage === totalPages;
        return (
            <StyledDiv>
                <Grid padding="10px">
                    <GridCol center colXS={ 6 } colMD={ 4 }>
                        <StyledSpan>Rows per page:</StyledSpan>
                    </GridCol>
                    <GridCol center colXS={ 6 } colMD={ 2 }>
                        <SelectField
                            width="80px"
                            options={ basicRowCount }
                            handleChange={ handleChangeRowsPerPage }
                            value={ rowsPerPage }
                        />
                    </GridCol>
                    <GridCol center colXS={ 12 } colMD={ 2 }>
                        <StyledSpan>
                            {currentPage}
                            {' '}
of
                            {' '}
                            {totalPages}
                        </StyledSpan>
                    </GridCol>
                    <GridCol center colXS={ 12 } colMD={ 4 }>
                        <StyledFaAngleDoubleLeft onClick={ handleChange(1) } disabled={ isFirst } />
                        <StyledAngleLeft onClick={ handleChange(dec(currentPage)) } disabled={ isFirst } />
                        <StyledAngleRight onClick={ handleChange(inc(currentPage)) } disabled={ isLast } />
                        <StyledFaAngleDoubleRight onClick={ handleChange(totalPages) } disabled={ isLast } />
                    </GridCol>
                </Grid>
            </StyledDiv>
        );
    }
}

TablePagination.propTypes = {
    /**
     * Number representing total number of table rows
     */
    totalRows: PropTypes.number,
    /**
     * Number of page we are currently at
     */
    currentPage: PropTypes.number,
    /**
     * Number of total available pages
     */
    totalPages: PropTypes.number,
    /**
     * Function to handle numberOfRows per page change
     */
    handleChangeRowsPerPage: PropTypes.func,
    /**
     * Function to handle page change, expecting (value) => (e) => {}
     */
    handleChangePage: PropTypes.func,
    /**
     * Number of rows per page
     */
    rowsPerPage: PropTypes.number,
};

export { StyledDiv };
export default TablePagination;
