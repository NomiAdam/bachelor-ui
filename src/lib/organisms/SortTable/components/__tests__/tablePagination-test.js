/* eslint-disable no-undef */
import React from 'react';
import {shallow} from 'enzyme';
import TablePagination, { StyledDiv } from '../TablePagination';

const mock = jest.fn();

describe('TablePagination', () => {
    beforeEach(() => {
        jest.resetAllMocks();
    });
    it('should be defined', () => {
        expect(shallow(<TablePagination/>)).toBeDefined();
    });
    it('should match the snapshot', () => {
        expect(shallow(<TablePagination/>)).toMatchSnapshot();
    });
    it('should be of type StyledDiv', () => {
        expect(shallow(<TablePagination/>).type()).toBe(StyledDiv);
    });
    it('should not trigger any mock', () => {
        shallow(<TablePagination
            handleChangeRowsPerPage={mock}
            handleChangePage={mock}
        />);
        expect(mock).not.toHaveBeenCalled();
    });
});
