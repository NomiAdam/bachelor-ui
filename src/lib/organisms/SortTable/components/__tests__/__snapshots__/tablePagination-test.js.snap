// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TablePagination should match the snapshot 1`] = `
<styled.div>
  <styled.div
    gridGap="10px"
    padding="10px"
  >
    <styled.div
      center={true}
      colMD={4}
      colXS={6}
    >
      <styled.span>
        Rows per page:
      </styled.span>
    </styled.div>
    <styled.div
      center={true}
      colMD={2}
      colXS={6}
    >
      <SelectField
        options={
          Array [
            Object {
              "label": 10,
              "value": 10,
            },
            Object {
              "label": 25,
              "value": 25,
            },
            Object {
              "label": 50,
              "value": 50,
            },
          ]
        }
        placeholder=""
        value=""
        width="80px"
      />
    </styled.div>
    <styled.div
      center={true}
      colMD={2}
      colXS={12}
    >
      <styled.span>
         
        of
         
        NaN
      </styled.span>
    </styled.div>
    <styled.div
      center={true}
      colMD={4}
      colXS={12}
    >
      <Styled(FaAngleDoubleLeft)
        disabled={false}
        onClick={[Function]}
      />
      <Styled(FaAngleLeft)
        disabled={false}
        onClick={[Function]}
      />
      <Styled(FaAngleRight)
        disabled={false}
        onClick={[Function]}
      />
      <Styled(FaAngleDoubleRight)
        disabled={false}
        onClick={[Function]}
      />
    </styled.div>
  </styled.div>
</styled.div>
`;
